// react-native-navigation/lib/dist/interfaces/Options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var OptionsModalPresentationStyle;
(function (OptionsModalPresentationStyle) {
    OptionsModalPresentationStyle[OptionsModalPresentationStyle["formSheet"] = 0] = "formSheet";
    OptionsModalPresentationStyle[OptionsModalPresentationStyle["pageSheet"] = 1] = "pageSheet";
    OptionsModalPresentationStyle[OptionsModalPresentationStyle["overFullScreen"] = 2] = "overFullScreen";
    OptionsModalPresentationStyle[OptionsModalPresentationStyle["overCurrentContext"] = 3] = "overCurrentContext";
    OptionsModalPresentationStyle[OptionsModalPresentationStyle["currentContext"] = 4] = "currentContext";
    OptionsModalPresentationStyle[OptionsModalPresentationStyle["popOver"] = 5] = "popOver";
    OptionsModalPresentationStyle[OptionsModalPresentationStyle["fullScreen"] = 6] = "fullScreen";
    OptionsModalPresentationStyle[OptionsModalPresentationStyle["none"] = 7] = "none";
})(OptionsModalPresentationStyle = exports.OptionsModalPresentationStyle || (exports.OptionsModalPresentationStyle = {}));
